Add Nunchaku Lite single-file quantization (#14100)
huggingface/diffusers: docs/source/en/quantization/nunchaku.md

<!--Copyright 2026 The HuggingFace Team. All rights reserved.

Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
the License. You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.

-->

# Nunchaku Lite

> Nunchaku Lite is a quantization backend for loading prequantized checkpoints in Diffusers. Use

Reviewer: Should provide some citations to the original Nunchaku repo since Nunchaku Lite is based off of it?

Author: Not only the repo, I will reference their paper too

Author: this is resolved

[`diffuse-compressor`](https://github.com/rootonchair/diffuse-compressor) to create compatible checkpoints by choosing
or adapting a target config for the model architecture, quantizing and exporting the transformer, and packaging the
result as a Diffusers pipeline repository.

## Load a quantized pipeline

Load the packaged prequantized pipeline with [`~DiffusionPipeline.from_pretrained`]. Diffusers reads the quantization
config from `config.json`.

```python
import torch
from diffusers import DiffusionPipeline

model_id = "rootonchair/ERNIE-Image-Turbo-nunchaku-lite-nvfp4"

pipe = DiffusionPipeline.from_pretrained(
    model_id,
    torch_dtype=torch.bfloat16,
)

> pipe.to("cuda")

Reviewer: ```suggestion
pipe = DiffusionPipeline.from_pretrained(
    model_id, torch_dtype=torch.bfloat16,
).to("cuda")
```

prompt = "A modern red armchair in a quiet studio, soft window light, realistic product photography"
image = pipe(
    prompt=prompt,
    height=1024,
    width=1024,
    num_inference_steps=8,
    guidance_scale=1.0,
    use_pe=False,
).images[0]
image.save("ernie-image-turbo-nunchaku-lite.png")

> ```

Reviewer: Is it compatible with offloading, too?

Author: yes it is compatible with `enable_sequential_cpu_offload` and `enable_model_cpu_offload`

## Using Optimized CUDA Kernels with Nunchaku Lite

Nunchaku Lite uses optimized CUDA kernels through the `kernels` package. The kernels are loaded automatically when you
load a Nunchaku Lite checkpoint. Install the runtime dependency before loading a Nunchaku Lite checkpoint.

```bash
pip install -U kernels
```

> [!NOTE]
> The exported state dict must match the target Diffusers model architecture exactly. For example, a checkpoint
> quantized with fused QKV projections won't load into a model config that expects separate Q, K, and V projection
> modules.

## Supported Quantization Types

Nunchaku Lite supports the following quantized linear layer formats.

> [!TIP]
> Use `fp4` on Blackwell GPUs. Running `int4` checkpoints on Blackwell can be slower than `fp4`.

| Method | Precision | Group size | Notes |
|---|---:|---:|---|
| `svdq_w4a4` | `fp4` | 16 | Uses NVFP4 runtime kernels with SVDQ low-rank correction. |
| `svdq_w4a4` | `int4` | 64 | Uses INT4 W4A4 kernels with SVDQ low-rank correction. |
| `awq_w4a16` | `int4` | 64 | Uses INT4 weight-only AWQ-style kernels. |

> 

Reviewer 2: maybe we should add some recommendation of hardware to be used with ?

Author: yes, I will be a good addition

Reviewer: I think it was not addressed? I think should mention a matrix of the quant schemes and their supported hardware / CUDA capabilities. Also, the note on hopper since we cannot support it.

## NunchakuLiteQuantizationConfig

> The `model.json` file must include a compact [`NunchakuLiteQuantizationConfig`]. It defines the runtime

Reviewer: Did you mean `config.json`?

`compute_dtype` and the target modules for each Nunchaku Lite quantization method.

- `compute_dtype`: runtime dtype for floating-point buffers in quantized modules, typically `torch.bfloat16`.
- `svdq_w4a4`: SVDQ W4A4 target config with `precision`, `group_size`, `rank`, and `targets`.
- `awq_w4a16`: AWQ W4A16 target config with `precision`, `group_size`, and `targets`.

The example below shows the expected shape with shortened target lists.

```json
{
  "_class_name": "ErnieImageTransformer2DModel",
  "quantization_config": {
    "quant_method": "nunchaku_lite",
    "compute_dtype": "bfloat16",
    "svdq_w4a4": {
      "precision": "fp4",
      "group_size": 16,
      "rank": 32,
      "targets": ["layers.0.self_attention.to_q"]
    },
    "awq_w4a16": {
      "precision": "int4",
      "group_size": 64,
      "targets": ["final_linear"]
    }
  }
}

> ```

Reviewer: Can we briefly summarize what would this config do? Would it apply a combination of `svdq_w4a4` and `awq_w4a16` to the `targets`?

Author: I have updated the details

## torch.compile

Nunchaku Lite kernels and quantized linear layers are compatible with [`torch.compile`](../optimization/fp16#torchcompile).
Compile the quantized transformer after loading the pipeline for faster inference.

```python

> pipe.transformer = torch.compile(pipe.transformer, mode="default", fullgraph=False)

Reviewer: Why can't we do it `fullgraph=True`? Can you open an issue on the nunchaku-lite repo with the trace?

Author: with the second kernel version, this has been resolved

<!--Copyright 2026 The HuggingFace Team. All rights reserved.

Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
the License. You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.

-->

# Nunchaku Lite

Nunchaku Lite is a quantization backend for loading prequantized checkpoints in Diffusers. Use
[`diffuse-compressor`](https://github.com/rootonchair/diffuse-compressor) to create compatible checkpoints by choosing
or adapting a target config for the model architecture, quantizing and exporting the transformer, and packaging the

> result as a Diffusers pipeline repository.

Reviewer 3: ```suggestion
Nunchaku Lite is a quantization backend for loading prequantized checkpoints in Diffusers. Create compatible checkpoints with [diffuse-compressor](https://github.com/rootonchair/diffuse-compressor). It quantizes and exports a transformer, then packages it as a Diffusers pipeline.
```

Nunchaku Lite builds on the original [Nunchaku](https://github.com/nunchaku-ai/nunchaku) inference engine,
[DeepCompressor](https://github.com/nunchaku-ai/deepcompressor) quantization library, and
[SVDQuant paper](https://arxiv.org/abs/2411.05007).

## Load a quantized pipeline

Load the packaged prequantized pipeline with [`~DiffusionPipeline.from_pretrained`]. Diffusers reads the quantization

> config from `config.json`.

Reviewer 3: ```suggestion
Load the prequantized pipeline with [`~DiffusionPipeline.from_pretrained`], which reads the quantization
config from `config.json`.
```

```python
import torch
from diffusers import DiffusionPipeline

model_id = "rootonchair/ERNIE-Image-Turbo-nunchaku-lite-nvfp4"

pipe = DiffusionPipeline.from_pretrained(
    model_id, torch_dtype=torch.bfloat16,
).to("cuda")

prompt = "A modern red armchair in a quiet studio, soft window light, realistic product photography"
image = pipe(
    prompt=prompt,
    height=1024,
    width=1024,
    num_inference_steps=8,
    guidance_scale=1.0,

>     use_pe=False,

Reviewer 3: may be helpful to either remove or very briefly explain `use_pe` is model-specific so users copying the snippet for another model won't get confused

Author: ```suggestion
```
let's remove it

).images[0]
image.save("ernie-image-turbo-nunchaku-lite.png")
```

> ## Using Optimized CUDA Kernels with Nunchaku Lite

Reviewer 3: ```suggestion
## Install the CUDA kernels
```

Nunchaku Lite uses optimized CUDA kernels through the `kernels` package. The kernels are loaded automatically when you

> load a Nunchaku Lite checkpoint. Install the runtime dependency before loading a Nunchaku Lite checkpoint.

Reviewer 3: ```suggestion
The kernels package supplies the optimized CUDA kernels, which load automatically. Install it first.
```

```bash
pip install -U kernels

> ```

Reviewer 3: move the installation before the first code snippet which needs it

> > [!NOTE]

Reviewer 3: would move this to the loading section where architecture matching is more relevant versus under CUDA kernels

> The exported state dict must match the target Diffusers model architecture exactly. For example, a checkpoint
> quantized with fused QKV projections won't load into a model config that expects separate Q, K, and V projection
> modules.

> ## Supported Quantization Types

Reviewer 3: ```suggestion
## Supported quantization types
```

Nunchaku Lite supports the following quantized linear layer formats.

> [!TIP]
> Use `nvfp4` on Blackwell GPUs. Running `int4` checkpoints on Blackwell can be slower than `nvfp4`.

The CUDA kernels currently support the following NVIDIA GPU architectures:

- `sm_75` (Turing, for example RTX 2080)
- `sm_80` (Ampere, for example A100)
- `sm_86` (Ampere, for example RTX 3090 and RTX A6000)
- `sm_89` (Ada, for example RTX 4090)
- `sm_120` (Blackwell, for example RTX 5090)

> [!NOTE]
> Hopper GPUs, such as `sm_90` H100 and H200, are not currently supported.

`nvfp4` checkpoints require a Blackwell or newer NVIDIA GPU. On Blackwell GPUs, use PyTorch >= 2.7 with CUDA >= 12.8.
`int4` checkpoints require a Turing or newer NVIDIA GPU.

| Method | Precision | Group size | Notes |
|---|---:|---:|---|
| `svdq_w4a4` | `nvfp4` | 16 | Uses NVFP4 runtime kernels with SVDQ low-rank correction. |
| `svdq_w4a4` | `int4` | 64 | Uses INT4 W4A4 kernels with SVDQ low-rank correction. |
| `awq_w4a16` | `int4` | 64 | Uses INT4 weight-only AWQ-style kernels. |

## NunchakuLiteQuantizationConfig

> The `config.json` file must include a compact [`NunchakuLiteQuantizationConfig`]. It defines the runtime

Reviewer 3: ```suggestion
The `config.json` file must include a [`NunchakuLiteQuantizationConfig`]. It defines the runtime
```

`compute_dtype` and the target modules for each Nunchaku Lite quantization method.

- `compute_dtype`: runtime dtype for floating-point buffers in quantized modules, typically `torch.bfloat16`.
- `svdq_w4a4`: SVDQ W4A4 target config with `precision`, `group_size`, `rank`, and `targets`.
- `awq_w4a16`: AWQ W4A16 target config with `precision`, `group_size`, and `targets`.

Each entry in `targets` must point to a linear layer. Diffusers replaces the linear layers listed under `svdq_w4a4`

> with SVDQ W4A4 layers and the linear layers listed under `awq_w4a16` with AWQ W4A16 layers. The example below shows the

Reviewer 3: ```suggestion
Each entry in `targets` must point to a linear layer. Diffusers swaps each `svdq_w4a4` target for an SVDQ W4A4 layer and each `awq_w4a16` target for an AWQ W4A16 layer. The example below shows the
```

expected shape with shortened target lists.

> [!NOTE]
> A target module can only use one Nunchaku Lite quantization method. Do not list the same target under both

> > `svdq_w4a4` and `awq_w4a16`.

Reviewer 3: ```suggestion
List each module you want to quantize under `svdq_w4a4` or `awq_w4a16`. A module can only use one method, so don't list the same target under both.
```

```json
{
  "_class_name": "ErnieImageTransformer2DModel",
  "quantization_config": {
    "quant_method": "nunchaku_lite",
    "compute_dtype": "bfloat16",
    "svdq_w4a4": {
      "precision": "nvfp4",
      "group_size": 16,
      "rank": 32,
      "targets": ["layers.0.self_attention.to_q"]
    },
    "awq_w4a16": {
      "precision": "int4",
      "group_size": 64,
      "targets": ["final_linear"]
    }
  }
}
```

## torch.compile

Nunchaku Lite kernels and quantized linear layers are compatible with [`torch.compile`](../optimization/fp16#torchcompile).
Compile the quantized transformer after loading the pipeline for faster inference.

```python
pipe.transformer = torch.compile(pipe.transformer, mode="default", fullgraph=True)
```

An ERNIE-Image-Turbo benchmark on an RTX PRO 6000 reported that Nunchaku Lite NVFP4 with `torch.compile` reduced the full pipeline latency from 2.271s to 1.675s. Compared to the original BF16 pipeline, the compiled

> Nunchaku Lite NVFP4 pipeline reached a 1.8x speedup.

Reviewer 3: front-load the benefits

```suggestion
The compiled Nunchaku Lite NVFP4 pipeline runs 1.8x faster than the original BF16 pipeline (2.271s → 1.675s on an RTX PRO 6000).
```
